lms/services/user.py

```diff
@@ -29,6 +29,7 @@ def upsert_user(self, lti_user: LTIUser) -> User:
             # Update the existing user from the fields which can change on a
             # new one
             existing_user.roles = new_user.roles
+            self._update_user(existing_user, lti_user)
             return existing_user
```

Comment:
Something seems a bit iffy here? I think we end up calling `_update_user` twice in the case of an existing user, which doesn't seem necessary.

This is quite hard to see as we use many out of method helpers. If you inline it all, then I think the structure is something like this:

```python
    def upsert_user(self, lti_user: LTIUser) -> User:
        """Store a record of having seen a particular user."""
        new_user =  User(
            application_instance_id=lti_user.application_instance_id,
            user_id=lti_user.user_id,
            roles=lti_user.roles,
            h_userid=lti_user.h_user.userid(self._h_authority),
        )
        self._update_user(new_user, lti_user)

        if existing_user := self._find_existing_user(model_user=new_user):
            # Update the existing user from the fields which can change on a
            # new one
            existing_user.roles = new_user.roles
            self._update_user(existing_user, lti_user)
            return existing_user

        self._db.add(new_user)
        return new_user
```

Whereas I think we could do something like:

```python
    def upsert_user(self, lti_user: LTIUser) -> User:
        """Store a record of having seen a particular user."""
        user = User(
            application_instance_id=lti_user.application_instance_id,
            user_id=lti_user.user_id,
            roles=lti_user.roles,
            h_userid=lti_user.h_user.userid(self._h_authority),
        )

        if existing_user := self._find_existing_user(model_user=user):
            # Update the existing user from the fields which can change on a
            # new one
            existing_user.roles = user.roles
            user = existing_user
        else:
            self._db.add(user)

        self._update_user(user, lti_user)

        return user
```

As the functions are only called once in-lining might be the way to go?

Reply:
Yeah that's nicer without the helpers :+1: